File: android/src/main/java/com/getcapacitor/community/Bitchat.java
package com.getcapacitor.community;

import static com.getcapacitor.community.BitchatHelper.makeUUID;

import android.content.Context;
import androidx.activity.ComponentActivity;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import com.bitchat.android.mesh.BluetoothMeshDelegate;
import com.bitchat.android.mesh.BluetoothMeshService;
import com.bitchat.android.model.BitchatMessage;
import com.bitchat.android.onboarding.BatteryOptimizationManager;
import com.bitchat.android.onboarding.PermissionManager;
import com.getcapacitor.community.classes.options.EstablishOptions;
import com.getcapacitor.community.classes.options.InitializeOptions;
import com.getcapacitor.community.classes.options.IsEstablishedOptions;
import com.getcapacitor.community.classes.options.SendOptions;
import com.getcapacitor.community.classes.options.StartOptions;
import com.getcapacitor.community.classes.results.EstablishResult;
import com.getcapacitor.community.classes.results.InitializeResult;
import com.getcapacitor.community.classes.results.IsEstablishedResult;
import com.getcapacitor.community.classes.results.IsInitializedResult;
import com.getcapacitor.community.classes.results.IsStartedResult;
import com.getcapacitor.community.classes.results.SendResult;
import com.getcapacitor.community.classes.results.StartResult;
import com.getcapacitor.community.interfaces.Callback;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;
import java.util.UUID;
import kotlin.Unit;
import org.jetbrains.annotations.NotNull;

public class Bitchat {

    private final String MISSING_PERMISSIONS = "missing permissions";
    private final String NOT_INITIALIZED = "not initialized";
    private final String NOT_STARTED = "not started";
    private final String MISSING_PAYLOAD = "missing payload";
    private final String MISSING_PEER_ID = "missing peer identifier";

    private boolean isInitialized = false;
    private boolean isStarted = false;

    private String nickname;

    @NonNull
    private final BitchatPlugin plugin;

    @NonNull
    private final BitchatConfig config;

    @NonNull
    private final BatteryOptimizationManager batteryOptimizationManager;

    @NonNull
    private final PermissionManager permissionManager;

    @NonNull
    private final BluetoothMeshService meshService;

    public Bitchat(@NonNull BitchatPlugin plugin, @NonNull BitchatConfig config) {
        this.plugin = plugin;
        this.config = config;

        ComponentActivity activity = plugin.getActivity();
        Context context = plugin.getContext();

        batteryOptimizationManager = new BatteryOptimizationManager(activity, context, () -> Unit.INSTANCE, (message) -> Unit.INSTANCE);

        // Initialize permission management
        permissionManager = new PermissionManager(context);
        // Initialize core mesh service first
        meshService = new BluetoothMeshService(context);
    }

    public boolean isBatteryOptimizationDisabled() {
        return batteryOptimizationManager.isBatteryOptimizationDisabled();
    }

    public void requestDisableBatteryOptimization() {
        batteryOptimizationManager.requestDisableBatteryOptimization();
    }

    /**
     * Initialize
     */

    public void initialize(@NonNull InitializeOptions options, @NonNull Callback callback) {
        meshService.setDelegate(
            new BluetoothMeshDelegate() {
                @Override
                public void didReceiveChannelLeave(@NonNull String channel, @NonNull String fromPeer) {
                    System.out.println("didReceiveChannelLeave: " + channel + ", " + fromPeer);
                }

                @Override
                public void didReceiveReadReceipt(@NonNull String messageID, @NonNull String recipientPeerID) {
                    System.out.println("didReceiveReadReceipt: " + messageID + ", " + recipientPeerID);
                }

                @Override
                public String getNickname() {
                    return nickname != null ? nickname : null;
                }

                @Override
                public void didReceiveMessage(@NonNull BitchatMessage message) {
                    String messageID = message.getId();
                    String content = message.getContent();
                    String peerID = message.getSenderPeerID();

                    Boolean isPrivate = message.isPrivate();
                    Boolean isRelay = message.isRelay();

                    plugin.onReceivedEvent(makeUUID(messageID), content, peerID, isPrivate, isRelay);
                }

                @Override
                public String decryptChannelMessage(@NonNull byte[] encryptedContent, @NonNull String channel) {
                    System.out.println("decryptChannelMessage: " + Arrays.toString(encryptedContent) + ", " + channel);
                    return "";
                }

                @Override
                public void didReceiveDeliveryAck(@NonNull String messageID, @NonNull String recipientPeerID) {
                    System.out.println("didReceiveDeliveryAck: " + messageID + ", " + recipientPeerID);
                }

                @Override
                public void didUpdatePeerList(@NonNull List<String> peers) {
                    plugin.onPeerListUpdatedEvent(peers);
                }

                @Override
                public boolean isFavorite(@NonNull String peerID) {
                    System.out.println("isFavorite: " + peerID);
                    return false;
                }

                @Override
                public void onStarted(@NonNull String peerID, @Nullable Boolean success) {
                    isStarted = success != null && success;

                    plugin.onStartedEvent(peerID, success);
                }

                @Override
                public void onStopped() {
                    isStarted = false;

                    plugin.onStoppedEvent();
                }

                @Override
                public void onFound(@NonNull String peerID, @NotNull String nickname) {
                    // establishNoiseSessionIfNeeded(peerID);

                    plugin.onFoundEvent(peerID, nickname);
                }

                @Override
                public void onLost(@NonNull String peerID) {
                    plugin.onLostEvent(peerID);
                }

                @Override
                public void onConnected(@NonNull String peerID) {
                    plugin.onConnectedEvent(peerID);
                }

                @Override
                public void onDisconnected(@NonNull String peerID) {
                    plugin.onDisconnectedEvent(peerID);
                }

                @Override
                public void onEstablished(@NotNull String peerID) {
                    plugin.onEstablishedEvent(peerID);
                }

                @Override
                public void onRSSIUpdated(@NonNull String peerID, int rssi) {
                    plugin.onRSSIUpdatedEvent(peerID, rssi);
                }

                @Override
                public void onPeerInfoUpdated(@NotNull String peerID, @NotNull String nickname, boolean isVerified) {
                    // plugin.onReceivedEvent(nickname, peerID);
                }

                @Override
                public void onPeerIDChanged(@NotNull String peerID, @Nullable String oldPeerID, @NotNull String nickname) {
                    plugin.onPeerIDChangedEvent(peerID, oldPeerID, nickname);
                }

                @Override
                public void onSent(@NonNull String messageID, @Nullable String peerID) {
                    plugin.onSentEvent(makeUUID(messageID), peerID);
                }
            }
        );

        @Nullable
        Long announceInterval = options.getAnnounceInterval() != null ? options.getAnnounceInterval() : config.getAnnounceInterval();

        if (announceInterval != null) {
            meshService.setAnnounceInterval(announceInterval);
        }

        isInitialized = true;

        String peerID = meshService.getMyPeerID();

        InitializeResult result = new InitializeResult(peerID);
        callback.success(result);
    }

    public void isInitialized(@NonNull Callback callback) {
        IsInitializedResult result = new IsInitializedResult(isInitialized);
        callback.success(result);
    }

    public void start(@NonNull StartOptions options, @NonNull Callback callback) {
        if (!isInitialized) {
            callback.error(new Exception(NOT_INITIALIZED));
            return;
        }

        if (!permissionManager.areAllPermissionsGranted()) {
            callback.error(new Exception(MISSING_PERMISSIONS));
            return;
        }

        @Nullable
        String message = options.getMessage();

        if (message != null) {
            nickname = message;
        }

        meshService.startServices();

        String peerID = meshService.getMyPeerID();

        StartResult result = new StartResult(peerID);
        callback.success(result);
    }

    public void isStarted(@NonNull Callback callback) {
        IsStartedResult result = new IsStartedResult(isStarted);
        callback.success(result);
    }

    public void stop(@NonNull Callback callback) {
        meshService.stopServices();

        callback.success();
    }

    /**
     * Session
     */

    public void establish(@NonNull EstablishOptions options, @NonNull Callback callback) {
        if (!isInitialized) {
            callback.error(new Exception(NOT_INITIALIZED));
            return;
        }
        if (!isStarted) {
            callback.error(new Exception(NOT_STARTED));
            return;
        }

        @Nullable
        String peerID = options.getPeerID();

        if (peerID == null) {
            callback.error(new Exception(MISSING_PEER_ID));
            return;
        }

        if (!meshService.hasEstablishedSession(peerID)) {
            meshService.sendAnnouncementToPeer(peerID);
            meshService.initiateNoiseHandshake(peerID);
        }

        Boolean isEstablished = meshService.hasEstablishedSession(peerID);

        EstablishResult result = new EstablishResult(isEstablished);
        callback.success(result);
    }

    public void isEstablished(@NonNull IsEstablishedOptions options, @NonNull Callback callback) {
        @Nullable
        String peerID = options.getPeerID();

        if (peerID == null) {
            callback.error(new Exception(MISSING_PEER_ID));
            return;
        }

        Boolean isEstablished = meshService.hasEstablishedSession(peerID);

        IsEstablishedResult result = new IsEstablishedResult(isEstablished);
        callback.success(result);
    }

    /**
     * Payload
     */

    public void send(@NonNull SendOptions options, @NonNull Callback callback) {
        if (!isInitialized) {
            callback.error(new Exception(NOT_INITIALIZED));
            return;
        }
        if (!isStarted) {
            callback.error(new Exception(NOT_STARTED));
            return;
        }

        @Nullable
        String message = options.getMessage();

        if (message == null) {
            callback.error(new Exception(MISSING_PAYLOAD));
            return;
        }

        UUID messageID = UUID.randomUUID();

        @Nullable
        String peerID = options.getPeerID();

        if (peerID == null) {
            meshService.sendMessage(message, Collections.emptyList(), null);
        } else {
            meshService.sendPrivateMessage(message, peerID, "unknown", messageID.toString());
        }

        SendResult result = new SendResult(messageID);
        callback.success(result);
    }
}

File: ios/Sources/BitchatPlugin/Bitchat.swift
import Foundation
import Capacitor
import CoreBluetooth

@objc public class Bitchat: NSObject, BitchatDelegate {
    private var isInitialized: Bool = false
    private var isStarted: Bool = false

    private var nickname: String?

    private let plugin: BitchatPlugin
    private let config: BitchatConfig?

    private let meshService: BLEService
    private let identityManager: SecureIdentityStateManagerProtocol

    init(plugin: BitchatPlugin, config: BitchatConfig?) {
        self.plugin = plugin
        self.config = config

        let keychain: KeychainManager = KeychainManager()
        identityManager = SecureIdentityStateManager(keychain)

        self.meshService = BLEService(keychain: keychain, identityManager: identityManager)

        super.init()
    }

    deinit {
        stop {_ in}
    }

    /**
     * Initialize
     */

    @objc public func initialize(_ options: InitializeOptions, completion: @escaping (Result?, Error?) -> Void) {
        meshService.delegate = self

        if let announceInterval = options.getAnnounceInterval() ?? config?.getAnnounceInterval() {
            meshService.announceInterval = TimeInterval(announceInterval)
        }

        do {
            isInitialized = true

            let peerID = meshService.myPeerID.id

            let result = InitializeResult(peerID)
            completion(result, nil)
        } catch {
            completion(nil, error)
        }
    }

    @objc public func isInitialized(completion: @escaping (Result?, Error?) -> Void) {
        let result = IsInitializedResult(isInitialized)
        completion(result, nil)
    }

    @objc public func start(_ options: StartOptions, completion: @escaping (Result?, Error?) -> Void) {
        if !isInitialized {
            return completion(nil, CustomError.notInitialized)
        }

        if let message = options.getMessage() {
            nickname = message
        }

        do {
            // Set nickname before starting services
            if let nickname {
                meshService.setNickname(nickname)
            }

            // Start mesh service immediately
            meshService.startServices()

            let peerID = meshService.myPeerID.id

            let result = StartResult(peerID)
            completion(result, nil)
        } catch {
            completion(nil, error)
        }
    }

    @objc public func isStarted(completion: @escaping (Result?, Error?) -> Void) {
        let result = IsStartedResult(isStarted)
        completion(result, nil)
    }

    @objc public func stop(completion: @escaping (Error?) -> Void) {
        do {
            meshService.stopServices()

            completion(nil)
        } catch {
            completion(error)
        }
    }

    /**
     * Session
     */

    @objc public func establish(_ options: EstablishOptions, completion: @escaping (Result?, Error?) -> Void) {
        if !isInitialized {
            return completion(nil, CustomError.notInitialized)
        }
        if !isStarted {
            return completion(nil, CustomError.notStarted)
        }

        guard let peerID = options.getPeerID() else {
            return completion(nil, CustomError.peerIDMissing)
        }

        do {
            if !(meshService.hasEstablishedSession(with: PeerID(str: peerID))) {
                meshService.triggerHandshake(with: PeerID(str: peerID))
            }

            let isEstablished = meshService.hasEstablishedSession(with: PeerID(str: peerID))

            let result = EstablishResult(isEstablished)
            completion(result, nil)
        } catch {
            completion(nil, error)
        }
    }

    @objc public func isEstablished(_ options: IsEstablishedOptions, completion: @escaping (Result?, Error?) -> Void) {
        guard let peerID = options.getPeerID() else {
            return completion(nil, CustomError.peerIDMissing)
        }

        let isEstablished = meshService.hasEstablishedSession(with: PeerID(str: peerID))

        let result = IsEstablishedResult(isEstablished)
        completion(result, nil)
    }

    /**
     * Payload
     */

    @objc public func send(_ options: SendOptions, completion: @escaping (Result?, Error?) -> Void) {
        if !isInitialized {
            return completion(nil, CustomError.notInitialized)
        }
        if !isStarted {
            return completion(nil, CustomError.notStarted)
        }

        guard let message = options.getMessage() else {
            return completion(nil, CustomError.payloadMissing)
        }

        do {
            let messageID = UUID()

            if let peerID = options.getPeerID() {
                meshService.sendMessage(message, to: PeerID(str: peerID), messageID: messageID.uuidString.lowercased())
            } else {
                meshService.sendMessage(message, messageID: messageID.uuidString.lowercased())
            }

            let result = SendResult(messageID)
            completion(result, nil)
        } catch {
            completion(nil, error)
        }
    }

    /**
     Delegate
     */

    // MARK: - Message Reception

    func didReceiveMessage(_ message: BitchatMessage) {
        // Route to appropriate handler
        if message.isPrivate {
            SecureLogger.debug("📥 handlePrivateMessage called for message from \(message.sender)", category: .session)
        } else {
            SecureLogger.debug("📥 handlePublicMessage called for message from \(message.sender)", category: .session)
        }

        let messageID = message.id
        let content = message.content
        let peerID = message.sender

        let isPrivate = message.isPrivate
        let isRelay = message.isRelay

        plugin.onReceivedEvent(messageID: Helper.makeUUID(messageID), content, peerID: peerID, isPrivate: isPrivate, isRelay: isRelay)
    }

    // MARK: - Peer Connection Events

    func didConnectToPeer(_ peerID: PeerID) {
        SecureLogger.debug("🤝 Peer connected: \(peerID)", category: .session)

        // Register ephemeral session with identity manager
        identityManager.registerEphemeralSession(peerID: peerID, handshakeState: .none)
    }

    func didDisconnectFromPeer(_ peerID: PeerID) {
        SecureLogger.debug("👋 Peer disconnected: \(peerID)", category: .session)

        // Remove ephemeral session from identity manager
        identityManager.removeEphemeralSession(peerID: peerID)
    }

    func didUpdatePeerList(_ peers: [PeerID]) {
    }

    func didUpdateBluetoothState(_ state: CBManagerState) {
    }

    func onStarted(_ peerID: PeerID, success: Bool?) {
        let peerID = peerID.id

        if let success {
            isStarted = success
        }

        plugin.onStartedEvent(peerID, isStarted: success)
    }

    func onStopped() {
        isStarted = false

        plugin.onStoppedEvent()
    }

    func onFound(_ peerID: PeerID, nickname: String) {
        let peerID = peerID.id

        plugin.onFoundEvent(peerID, message: nickname)
    }

    func onLost(_ peerID: PeerID) {
        let peerID = peerID.id

        plugin.onLostEvent(peerID)
    }

    func onConnected(_ peerID: PeerID) {
        let peerID = peerID.id

        plugin.onConnectedEvent(peerID)
    }

    func onDisconnected(_ peerID: PeerID) {
        let peerID = peerID.id

        plugin.onDisconnectedEvent(peerID)
    }

    func onEstablished(_ peerID: PeerID) {
        let peerID = peerID.id

        plugin.onEstablishedEvent(peerID)
    }

    func onSent(_ messageID: String, peerID: PeerID?) {
        guard let messageID = Helper.makeUUID(messageID) else { return }

        if let peerID = peerID {
            plugin.onSentEvent(messageID, peerID: peerID.id)
        } else {
            plugin.onSentEvent(messageID)
        }
    }

    func onRSSIUpdated(_ peerID: PeerID, rssi: Int) {
        let peerID = peerID.id

        plugin.onRSSIUpdatedEvent(peerID, rssi)
    }

    func onPeerInfoUpdated(_ peerID: PeerID, nickname: String, isVerified: Bool) {
    }

    func onPeerIDChanged(_ peerID: PeerID, oldPeerID: String?, nickname: String) {
    }

    /**
     * Permissions
     */

    private func bluetoothAuthorization() -> CBManagerAuthorization {
        if #available(iOS 13.1, *) {
            return CBCentralManager.authorization
        } else if #available(iOS 13.0, *) {
            return CBCentralManager().authorization
        }
    }

    private func bluetoothState() -> CBManagerState {
        return CBCentralManager().state
    }

    @objc public func checkPermissions(completion: @escaping (Result?, Error?) -> Void) {
        let bluetoothState = switch bluetoothAuthorization() {
        case .notDetermined:
            "prompt"
        case .restricted, .denied:
            "denied"
        case .allowedAlways:
            "granted"
        @unknown default:
            "prompt"
        }

        let result = PermissionsResult(bluetooth: bluetoothState)

        completion(result, nil)
    }

    @objc public func requestPermissions(_ options: RequestPermissionsOptions, completion: @escaping (Result?, Error?) -> Void) {
        var permissions = options.getPermissions() ?? []

        if permissions.isEmpty {
            permissions = ["bluetooth"]
        }

        let group = DispatchGroup()

        if permissions.contains("bluetooth") {
            if bluetoothState() == .poweredOff {
                guard let settingsUrl = URL(string: UIApplication.openSettingsURLString) else {
                    completion(nil, CustomError.openSettingsError)
                    return
                }

                DispatchQueue.main.async {
                    if UIApplication.shared.canOpenURL(settingsUrl) {
                        group.enter()

                        UIApplication.shared.open(settingsUrl, completionHandler: { (_) in
                            group.leave()
                        })
                    } else {
                        completion(nil, CustomError.openSettingsError)
                        return
                    }
                }
            }
        }

        group.notify(queue: DispatchQueue.main) {
            self.checkPermissions(completion: completion)
        }
    }
}
